pkg/operator/featureupgradablecontroller/feature_upgradeable_controller.go

```diff
@@ -0,0 +1,139 @@
+package featureupgradablecontroller
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/klog"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+var (
+	featureUpgradeableControllerWorkQueueKey = "key"
+
+	featureGatesAllowingUpgrade = sets.NewString("")
+)
+
+// FeatureUpgradeableController is a controller that sets upgradeable=false if anything outside the whitelist is the specified featuregates.
+type FeatureUpgradeableController struct {
+	operatorClient    v1helpers.OperatorClient
+	featureGateLister configlistersv1.FeatureGateLister
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+}
+
+func NewFeatureUpgradeableController(
+	operatorClient v1helpers.OperatorClient,
+	configInformer configinformers.SharedInformerFactory,
+	eventRecorder events.Recorder,
+) *FeatureUpgradeableController {
+	c := &FeatureUpgradeableController{
+
+		operatorClient:    operatorClient,
+		featureGateLister: configInformer.Config().V1().FeatureGates().Lister(),
+		eventRecorder:     eventRecorder.WithComponentSuffix("feature-upgradeable"),
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "FeatureUpgradeableController"),
+	}
+
+	operatorClient.Informer().AddEventHandler(c.eventHandler())
+	configInformer.Config().V1().FeatureGates().Informer().AddEventHandler(c.eventHandler())
+
+	c.cachesToSync = append(c.cachesToSync, operatorClient.Informer().HasSynced, configInformer.Config().V1().FeatureGates().Informer().HasSynced)
+	return c
+}
+
+func (c *FeatureUpgradeableController) sync() error {
+	featureGates, err := c.featureGateLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	cond := newUpgradeableCondition(featureGates)
+	if _, _, updateError := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(cond)); updateError != nil {
+		return updateError
+	}
+
+	return nil
+}
+
+func newUpgradeableCondition(featureGates *configv1.FeatureGate) operatorv1.OperatorCondition {
+	if featureGatesAllowingUpgrade.Has(string(featureGates.Spec.FeatureSet)) {
+		return operatorv1.OperatorCondition{
+			Type:   "FeatureGatesUpgradeable",
+			Reason: "AllowedFeatureGates_" + string(featureGates.Spec.FeatureSet),
```

Comment:
how will this look like as a string?

Reply:
> how will this look like as a string?

A little weird, but not distractingly bad.